### PR TITLE
Fix issue with single element through LazyFlow

### PR DESCRIPTION
**Pull request by aperkins1310**
## Purpose
This PR allows the LazyFlow to work with a single element, when the future is completed after the first element arrives. This previously worked in 2.5, but moving to 2.6 broke this for us, hence we realised the behaviour had changed.

## References
The code is reverted back to what it was before #26446, where this problem seems to have been introduced.

## Changes
- Don't complete promise if still switching
- Have a separate behaviour for the first element

## Background Context
Created failing test reproducing the issue, reverted back to the previously working version, and then eliminated differences until only the important changes remained.

### CODE REVIEW

**Comment by lightbend-cla-validator** [comment]
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

**Comment by aperkins1310** [comment]
Just to let you know, I have signed the CLA and have received a confirmation email, I only did it just before raising this though so assuming that’s why the check is failing

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: I think it would be better to capture `firstElement` in an `OptionVal` `var` here and use the presence of that as the flag that it needs to be pushed, and then toggle it to `OptionVal.None` after emitting it. That'd avoid  keeping a reference to `firstElement` for the rest ot the life of the inner flow (hindering it from getting GC:d)

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: So if upstream finishes  here and we have not yet seen a pull from the inner flow we cancel out the finish here and then complete when after pushing the element on first pull, after emitting the element. Makes sense. 👍

**Comment by johanandren** [comment]
OK TO TEST

**Comment by johanandren** [comment]
Looks like we had one CLA sign from you previously as well, from 2018. It's a bit weird that the CLA system allows more than one signature for the same handle, not sure if that is causing the validator to fail.

**Comment by johanandren** [comment]
I think it could also be that it needs the email associated with the CLA in the commit author metadata (comes from the git config, `git config --global user.email`). I think you can amend that to existing commits using `git commit --amend --author="Author Name <email@address.com>"` and then force push it to the branch.

**Comment by lightbend-cla-validator** [comment]
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala
Comment (aperkins1310) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: Done :)

**Comment by aperkins1310** [comment]
Got it now, I hadn't got my personal email attached to GitHub 👍

**Comment by johanandren** [comment]
CI is failing on formatting, if you run akka-stream/scalafmtAll and push the result that should fix it

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala
Comment (eyalfa) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: @johanandren , seems like reimplementing 'prepend' no?
what's wrong with leaving the prepend in place and completing the subOutlet anyway? this is the approach taken in `flatmapPrefix` where it's possible for the external graph stage to complete while the internal flow keeps running.

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala
Comment (eyalfa) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: this looks like a reimplementation of 'prepend' and I'm pretty sure it's not necessary here.
I propose a much simpler approach: when subOutlet is pulled (by prepended internal flow) it delegates to 'in', when 'in' is completed/failed it delegates to subOutlet. remember the internal flow has its own behaviour for dealing with completion (i.e. concat, orElse, prepend...) and errors (i.e. recover) so you have to make sure to delegate events to the internal prepended flow so it can take care of them.

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/LazyFlowSpec.scala
Comment (eyalfa) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/LazyFlowSpec.scala]: this races with data flow in the stream, you have no guarantee what happens first, is it the test completing the future or the LazyFlow consuming the first element.
I think u need more control here over the order of execution, either by introducing a second promise that the flow factory completes in order to signal the test code that the factory is invoked by the graph stage, or by other means (use a probe for source...).

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala
Comment (johanandren) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: Sounds like a good idea, I didn't consider that.

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/LazyFlowSpec.scala
Comment (aperkins1310) [reply akka-stream-tests/src/test/scala/akka/stream/scaladsl/LazyFlowSpec.scala]: I'm happy to change this one, you're right in there being a race condition. Should I update the other existing 'after the fact' tests here too? They're all susceptible to this same race condition

**Review by jrudolph — COMMENTED** [review]
Heretic question, but could we just remove the stage implementation and replace it by a usage of `flatMapPrefix`?

**Comment by jrudolph** [comment]
> Heretic question, but could we just remove the stage implementation and replace it by a usage of `flatMapPrefix`?

To answer myself: currently not possible because `flatMapPrefix` doesn't allow control over the materialized value (but that might be something we'd like to anyway?).

**Comment by eyalfa** [comment]
> > Heretic question, but could we just remove the stage implementation and replace it by a usage of `flatMapPrefix`?
> 
> To answer myself: currently not possible because `flatMapPrefix` doesn't allow control over the materialized value (but that might be something we'd like to anyway?).

what do u mean by 'control'? flatMapPrefix currently lacks an async version (where f returns a Future[Flow] instead of a Flow), however I'd rather not implement such an API as a graph stage but rather rely on a solid 'futureFlow' implementation. this way we'd have to stages handling different concerns that can be combined together to introduce elaborate constructs.

**Comment by eyalfa** [comment]
@aperkins1310 , I'll have a look at your last commits later today.
however, when browsing the current implementation it seems to me that the root cause for the behaviour you're experiencing is in the way onUpstreamFinished/Failure behaves when the switch is in motion:
```
override def onUpstreamFinish(): Unit = {
        if (!matPromise.isCompleted)
          matPromise.tryFailure(new NeverMaterializedException)
        // ignore onUpstreamFinish while the stage is switching but setKeepGoing
        if (switching) {
          setKeepGoing(true)
        } else {
          super.onUpstreamFinish()
        }
      }
```
notice there's an inherent race condition here as this signal (and the failure as well) might come out of band and definitely may 'beat' the flow future's completion.
I think once the switch is in motion, these signals should be stashed away and handed over to the future flow once the future completes. I'd further say the same for the downstream signals, see flatMapPrefix implementation and tests for reference, these principles were actually requested by @jrudolph when reviewing #28380 .

**Comment by jrudolph** [comment]
> what do u mean by 'control'? flatMapPrefix currently lacks an async version (where f returns a Future[Flow] instead of a Flow), however I'd rather not implement such an API as a graph stage but rather rely on a solid 'futureFlow' implementation. this way we'd have to stages handling different concerns that can be combined together to introduce elaborate constructs.

I agree, there's already `Flow.future` that could be used as a building block, the problem is that there's currently no way with `Flow.flatMapPrefix` to promote the materialized value to the outside.

**Comment by jrudolph** [comment]
>  the problem is that there's currently no way with `Flow.flatMapPrefix` to promote the materialized value to the outside.

Added https://github.com/akka/akka/issues/28608 to track adding that.

**Comment by eyalfa** [comment]
> > what do u mean by 'control'? flatMapPrefix currently lacks an async version (where f returns a Future[Flow] instead of a Flow), however I'd rather not implement such an API as a graph stage but rather rely on a solid 'futureFlow' implementation. this way we'd have to stages handling different concerns that can be combined together to introduce elaborate constructs.
> 
> I agree, there's already `Flow.future` that could be used as a building block, the problem is that there's currently no way with `Flow.flatMapPrefix` to promote the materialized value to the outside.

actually [there is] 😄

**Comment by eyalfa** [comment]
@jr

> > what do u mean by 'control'? flatMapPrefix currently lacks an async version (where f returns a Future[Flow] instead of a Flow), however I'd rather not implement such an API as a graph stage but rather rely on a solid 'futureFlow' implementation. this way we'd have to stages handling different concerns that can be combined together to introduce elaborate constructs.
> 
> I agree, there's already `Flow.future` that could be used as a building block, the problem is that there's currently no way with `Flow.flatMapPrefix` to promote the materialized value to the outside.

Thing is `Flow.future` is implemented in terms of `Flow.lazyFutureFlow` which gets us to this 😄 , I think it'd be nice to refactor `futureFlow` into its own graph stage (no conditions involved) and then implement the other variants in terms of `flamMapPrefix` and `futureFlow`

**Comment by jrudolph** [comment]
Yes :) Sorry about that really confusing conversation.

**Comment by eyalfa** [comment]
> @jrudolph
> 
> > > what do u mean by 'control'? flatMapPrefix currently lacks an async version (where f returns a Future[Flow] instead of a Flow), however I'd rather not implement such an API as a graph stage but rather rely on a solid 'futureFlow' implementation. this way we'd have to stages handling different concerns that can be combined together to introduce elaborate constructs.
> > 
> > 
> > I agree, there's already `Flow.future` that could be used as a building block, the problem is that there's currently no way with `Flow.flatMapPrefix` to promote the materialized value to the outside.
> 
> Thing is `Flow.future` is implemented in terms of `Flow.lazyFutureFlow` which gets us to this 😄 , I think it'd be nice to refactor `futureFlow` into its own graph stage (no conditions involved) and then implement the other variants in terms of `flamMapPrefix` and `futureFlow`

🖕 @jrudolph , do u think this is the right direction? if so we should open an issue describing it

**Comment by jrudolph** [comment]
> Thing is `Flow.future` is implemented in terms of `Flow.lazyFutureFlow` which gets us to this smile , I think it'd be nice to refactor `futureFlow` into its own graph stage (no conditions involved) and then implement the other variants in terms of `flamMapPrefix` and `futureFlow`

Thinking a bit more about this on the weekend, I wonder if it wouldn't be better after all to make the `flatMapPrefix` implementation Future-aware. For that stage it seems like a somewhat small change. The main reason would be that stages like `flatMapPrefix` or the proposed `futureFlow` stage need the dynamic stream creation with subsinks and subsources that introduce multiple levels of indirection that can not be easily optimized away. Since `flatMapPrefix` already has the indirection built in, adding Future support would only mean deferring to run the stream but at the same runtime cost per element.

To explain the indirection cost of subsource/subsink:
 * the dynamic substream runs in a separate GraphInterpreter that is attached to the same actor
 * each call crossing SubSink/SubSource goes through an async callback that is shortcut on the actor level to dispatch execution on the sibling GraphInterpreter
 * for a dynamic Flow that is dynamically connected on both ends there are basically two such more expensive handovers per element

WDYT?

**Comment by eyalfa** [comment]
> > Thing is `Flow.future` is implemented in terms of `Flow.lazyFutureFlow` which gets us to this smile , I think it'd be nice to refactor `futureFlow` into its own graph stage (no conditions involved) and then implement the other variants in terms of `flamMapPrefix` and `futureFlow`
> 
> Thinking a bit more about this on the weekend, I wonder if it wouldn't be better after all to make the `flatMapPrefix` implementation Future-aware. For that stage it seems like a somewhat small change. The main reason would be that stages like `flatMapPrefix` or the proposed `futureFlow` stage need the dynamic stream creation with subsinks and subsources that introduce multiple levels of indirection that can not be easily optimized away. Since `flatMapPrefix` already has the indirection built in, adding Future support would only mean deferring to run the stream but at the same runtime cost per element.
> 
> To explain the indirection cost of subsource/subsink:
> 
> * the dynamic substream runs in a separate GraphInterpreter that is attached to the same actor
> * each call crossing SubSink/SubSource goes through an async callback that is shortcut on the actor level to dispatch execution on the sibling GraphInterpreter
> * for a dynamic Flow that is dynamically connected on both ends there are basically two such more expensive handovers per element
> 
> WDYT?

I'd have to take a deeper dive in the code but I have a recollection of using the sub fusing materializer to materialize the nested flow for flatMapPrefix so I don't think it pays any async boundaries cost at all.
if I'm right then the same technique can be used in order to implement futureFlow.

my main concern with this approach is complicating the flatMapPrefix implementation in order to achieve something which can be easily achieved by composition, my tendency purely for design considerations is having a separate implementation for flatMapPrefix and futureFlow as each present the simplest possible solution to a distinct problem/challenge.

I think your concerns express reality where sometimes purity has to move aside for optimizations...

**Comment by jrudolph** [comment]
> I'd have to take a deeper dive in the code but I have a recollection of using the sub fusing materializer to materialize the nested flow for flatMapPrefix so I don't think it pays any async boundaries cost at all.
> if I'm right then the same technique can be used in order to implement futureFlow.

Yes, the actual async dispatch is optimized away but there's still a whole bunch of additional dispatch infrastructure that needs to run per element.

> my main concern with this approach is complicating the flatMapPrefix implementation in order to achieve something which can be easily achieved by composition, my tendency purely for design considerations is having a separate implementation for flatMapPrefix and futureFlow as each present the simplest possible solution to a distinct problem/challenge.
>
> I think your concerns express reality where sometimes purity has to move aside for optimizations...

I agree, it's a hard call. Maybe a bit of experimentation of the different alternatives could help.

**Comment by eyalfa** [comment]
@jrudolph , care to move this discussion to an issue? seems like we 'took a ride' on this PR for a much broader discussion

**Comment by jrudolph** [comment]
> @jrudolph , care to move this discussion to an issue? seems like we 'took a ride' on this PR for a much broader discussion

Sure, sorry, wasn't even aware I was still derailing this discussion.

**Review by jrudolph — APPROVED** [review]
This one LGTM but it seems it is broken for zero elements (just passes completion through without trying to materialize the lazy future).

**Comment by aperkins1310** [comment]
Sorry guys a lot of what you were saying went straight over my head so I took a step back on this.

@jrudolph can you explain a little bit further on this:
> This one LGTM but it seems it is broken for zero elements (just passes completion through without trying to materialize the lazy future).

I'm not quite sure what the expected behaviour is if it isn't this^ ? If there's zero elements I would've thought the behaviour would be "complete without creating internal flow when there was no elements in the stream" as the test states?

**Review by seglo — DISMISSED** [review]
LGTM. I don't understand either why we should care about materializing the future flow if an element is never pushed, but I'll wait for @jrudolph to respond before merging.

**Review by seglo — COMMENTED** [review]
Undoing this approval for now after reviewing some related `LazyFlow` cases about how to handle exceptions with inner flows consistently.  As this comment mentions https://github.com/akka/akka/issues/28016#issuecomment-595059713 we should also aim to have consistency about wrapping the downstream exception in a `NeverMaterializedExceptions` if the inner flow is never materialized (such as when no elements are pushed to the inner flow).

**Comment by seglo** [comment]
I think this PR should be closed as the behaviour @aperkins1310 has now been implemented as part of #28729 satisfies the new test that was added in this PR.

**Comment by seglo** [comment]
This is superseded with #28729 that's now merged. Thanks for identifying the issue.

**Pull request closed without merging**
